src/Microsoft.Azure.SignalR.Protocols/MulticastDataMessage.cs

```diff
@@ -198,4 +198,101 @@ public MultiGroupBroadcastDataMessage(IReadOnlyList<string> groupList, IDictiona
             GroupList = groupList;
         }
     }
-}
+
+    /// <summary>
+    /// A data message to indicate a client invocation request.
+    /// </summary>
+    public class ClientInvocationMessage : MulticastDataMessage
+    {
+        /// <summary>
+        /// Initialize a new instance of <see cref="ClientInvocationMessage"/> class.
+        /// </summary>
+        /// <param name="invocationId">The Id of the client invocation request.</param>
+        /// <param name="connectionId">The client connection Id.</param>
+        /// <param name="callerId">The serverId that init the client invocation.</param>
+        /// <param name="payloads">The payload of the message.</param>
+        /// <param name="tracingId">The tracing Id of the message.</param>
+        public ClientInvocationMessage(string invocationId, string connectionId, string callerId, IDictionary<string, ReadOnlyMemory<byte>> payloads, ulong? tracingId = null)
+            : base(payloads, tracingId)
+        {
+            InvocationId = invocationId;
+            ConnectionId = connectionId;
+            CallerId = callerId;
+        }
+
+        /// <summary>
+        /// Gets or sets the Id of the client invocation request.
+        /// </summary>
+        public string InvocationId { get; set; }
+
+        /// <summary>
+        /// Gets or sets the client connection Id.
+        /// </summary>
+        public string ConnectionId { get; set; }
+
+        /// <summary>
+        /// Gets or sets the caller server Id that init the client invocation.
+        /// </summary>
+        public string CallerId { get; set; }
+    }
+
+    /// <summary>
+    /// A data message of client invocation completion result.
+    /// </summary>
+    public class ServiceCompletionMessage : MulticastDataMessage
```

Comment:
completion message is always to that specific caller, why it is MulticastDataMessage?

Reply:
Considering it won't happen in handshake periods and route server always knows the protocol, make it kind of `ConnectionMessage` is better.